### PR TITLE
Overhaul ORA remote's HTTP access to RIA stores

### LINKED ISSUE #5451: enableremote ORA doesn't fail on invalid/inaccessible HTTP URL
In opposition to `ria+(ssh|file):// ...` URLs, ORA's `initremote` method doesn't properly test it can actually find a RIA store at the given HTTP URL. This in turn leads to `clone` not reconfiguring, when cloning via a different scheme.

### PR DESCRIPTION
**Pull request by bpoldrack**
Restructure ORA remote's `HTTPRemoteIO` to remove special cases in ORA's code, where HTTP is treated somewhat differrent than SSH or local access.
For RIA stores, this leads to the `ria-layout-version` files being required to be served as well (which technically wasn't the case before). It also means that the request URLs now use paths for annexed objects according to the specified layout, whereas before we'd always send the same request and it was up to the server to consider its layout and deliver the right thing.

HTTP stores are validated on `annex initremote`/`annex enableremote` the same way they are when using `ssh:` or `file:` scheme URLs. Therefore, autoenabling will fail under the same conditions, making `clone`'s reconfiguration business consistent across access methods.

Closes #5451

### CODE REVIEW

File: datalad/distributed/ora_remote.py
Comment (kyleam) [inline datalad/distributed/ora_remote.py:1049]: * It looks like `.exists` will fail if this isn't a pathlib object

 * Is it safe to assume the bare repo `HEAD` exists in a store?  It won't with, e.g., stores created with `create-sibling-ria --storage-sibling=only`, right?

Comment (bpoldrack) [reply datalad/distributed/ora_remote.py:1049]: > Is it safe to assume the bare repo HEAD exists in a store? It won't with, e.g., stores created with create-sibling-ria --storage-sibling=only, right?

Good point, forgot we allow for the absence of a git repo. All that is actually required for the remote to work is correct response to requests (HEAD|GET) to annex keys. But for initremote/enableremote there's no key we could generally assume to be there. So, I wonder what to test for. We could execute a checkpresent for an arbitrary key. Issue is: What does a 404 really tell us? We can tell host was resolved and there's a webserver responding. That's at least something. On the other hand that doesn't verify we can get anything related to that dataset.
May be we need to just define something else, a valid RIA store needs to serve over HTTP. `ria-layout-version` (at least at dataset level, not necessarily top-level) would be an obvious choice that would also increase flexibility for URL building (mapping to a different layout could then be done client-side just like with SSH/local). What do you think, @mih?

Comment (mih) [reply datalad/distributed/ora_remote.py:1049]: I hope I was not blocking this PR for 5 months, because I have nothing substantial to say. #5459 is rather terse and does not have any information on the context of the error. Taking what is provided it is unclear to me what `initremote` could do to prevent a future `enableremote` from failing. A server could vanish any time, or become misconfigured. To me this sounds like a UX of `enableremote` rather than a problem of `initremote`. Of course it would be good to check sanity of URLs given to `initremote`.

Looking at the code at the code I see that such checks are made in ~`get_store()`, but are explicitly disabled for HTTP access in `initremote()`. I'd say that it needs no change in behavior or logic, just a removal of the shortcuts made for HTTP access.

Comment (bpoldrack) [reply datalad/distributed/ora_remote.py:1049]: @mih : There's no difference between `initremote` and `enableremote` from the point of view of a special remote. `INITREMOTE` is what it gets from annex it both cases.

> Looking at the code at the code I see that such checks are made in ~get_store(), but are explicitly disabled for HTTP access in initremote(). I'd say that it needs no change in behavior or logic, just a removal of the shortcuts made for HTTP access.

Yes and no. Needs an equivalent. But in opposition to local/SSH access we don't really need the same things to be checked, since the special remote doesn't actually operate on the store's FS tree.

I will change this to request the `ria-layout-version` files both from the store- and dataset level. This will be a more general change, as ORA will then use that file to determine the key URL, just like in the case of local FS and SSH. So, abandon the idea that we approach it differently via HTTP in that we only ever send uniform requests independent of any `version` f the store and leave it to the server to map the requests differently, if needed. Instead we get that version and build the URL from it exactly like we'd do it for non-HTTP access.
This not only gives us something to verify but also enables another change in the future: A version file, that doesn't necessarily give a number/label referring to a predefined-in-datalad layout, but alternatively specifies the layout directly, so that a store's maintainer has all the flexibility to set it up w/o a need for the special remote to know the layout. I wanted to go there anyway, so I think this is going to solve more than one problem.

**Comment by bpoldrack** [comment]
Changed the approach and updated PR description accordingly.

File: datalad/distributed/ora_remote.py
Comment (mih) [reply datalad/distributed/ora_remote.py]: I am resolving this old conversation as it seems no longer relevant.

File: datalad/distributed/ora_remote.py
Comment (mih) [inline datalad/distributed/ora_remote.py:639]: Unrelated to this PR, but if downstream code requires a URL to not have a trailing slash, this better be an unconditional `.rstrip('/')`.

File: datalad/distributed/ora_remote.py
Comment (mih) [inline datalad/distributed/ora_remote.py:661]: `str()` is redundant. `as_posix()` returns a `str`.

File: datalad/distributed/ora_remote.py
Comment (mih) [inline datalad/distributed/ora_remote.py:672]: Another redundant `str()`

File: datalad/distributed/ora_remote.py
Comment (mih) [inline datalad/distributed/ora_remote.py:1117]:
```suggestion
                # TODO: That construction of "http(s)://host/" should probably be
```

**Review by mih — CHANGES_REQUESTED** [review]
Thanks for the changes. I left a few minor comments.

There should be a test though. Given that the original issue is about not failing, in case of an invallid URL such a test requires no particular infrastructure, just a random URL with no RIA store behind it.

**Review by mih — APPROVED** [review]
LGTM. Thanks for the changes.

**Pull request merged**
